### PR TITLE
Added KDTree method to generate cfs

### PR DESCRIPTION
**Pull request by soundarya98**
1. Added support to return k closest points from the dataset using a KD Tree
2. Added method in public_data_interface.py to predict the model on the entire dataset (required for KD tree implementation)
3. Added method in public_data_interface.py to predict the model on append the predictions to a copy of the dataset (required for KD tree implementation)

### CODE REVIEW

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        self.cfs = []
+        self.features_to_vary = []
+        self.cf_init_weights = []  # total_CFs, algorithm, features_to_vary
+        self.loss_weights = []  # yloss_type, diversity_loss_type, feature_weights
+        self.feature_weights_input = ''
+        self.hyperparameters = [1, 1, 1]  # proximity_weight, diversity_weight, categorical_penalty
+        self.optimizer_weights = []  # optimizer, learning_rate
+
+        # number of output nodes of ML model
+        temp_input = tf.convert_to_tensor([tf.random.uniform([len(self.data_interface.encoded_feature_names)])], dtype=tf.float32)
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:43]: can this line be written without explicitly invoking tensorflow? One option is to implement a function called `provide_temp_input` inside the model class `self.model.`

That way, this same code can work for all types of models and you don't need to write a different explainer for tensorflow, sklearn etc. The model interface will implement any model specific method.

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        dataset_dict_output = np.array([dataset_dict.values], dtype=np.float32)
+
+        self.predictions = self.predict_fn(dataset_dict_output)
+        self.predictions_vals = np.reshape(self.predictions[0], (self.predictions.shape[1],))
+
+        #segmenting the dataset according to outcome
+        #TODO: automate this for each outcome
+        dataset_with_predictions_all = self.data_interface.append_predictions(self.predictions_vals)
+        self.dataset_with_predictions = {}
+        self.dataset_with_predictions[1] = dataset_with_predictions_all.loc[np.round(dataset_with_predictions_all[self.data_interface.outcome_name + "_pred"])==1]
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:58]: note somewhere in the documentation of the class  that this works only for binary outcomes

File: dice_ml/data_interfaces/public_data_interface.py [hunk 3]
```diff
@@ -356,6 +356,18 @@ def prepare_query_instance(self, query_instance, encode):
 
             return temp.tail(test.shape[0]).reset_index(drop=True)
 
+    def prepare_dataset_instance(self, encode=True):
+        """Prepares the entire dataset as a sequence of test inputs for DiCE."""
+        dataset_dict = self.data_df[self.feature_names].to_dict('records')
```
Comment (amit-sharma) [inline dice_ml/data_interfaces/public_data_interface.py:361]: this is a bit odd. Line 361 converts df to dict, and then 362 converts it back to dict. Is there a better way to do this?
One way will be to break up the prepare query instance function into two, and only call the second part.

File: dice_ml/data_interfaces/public_data_interface.py [hunk 4]
```diff
@@ -356,6 +356,18 @@ def prepare_query_instance(self, query_instance, encode):
...
 
+    def prepare_dataset_instance(self, encode=True):
+        """Prepares the entire dataset as a sequence of test inputs for DiCE."""
+        dataset_dict = self.data_df[self.feature_names].to_dict('records')
+        return self.prepare_query_instance(query_instance=dataset_dict, encode=encode)
+
+    def append_predictions(self, test_pred):
+        """Appends the value predicted for each row of the dataset to a copy of the dataset"""
+        df_copy = self.data_df.copy()
+        for i in range(len(test_pred)):
```
Comment (amit-sharma) [inline dice_ml/data_interfaces/public_data_interface.py:367]: does this need a loop? would `df_copy["col_name"]=test_pred` work?

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        self.model = model_interface
+
+        # loading trained model
+        self.model.load_model()
+
+        # variables required to generate CFs - see generate_counterfactuals() for more info
+        self.cfs = []
+        self.features_to_vary = []
+        self.cf_init_weights = []  # total_CFs, algorithm, features_to_vary
+        self.loss_weights = []  # yloss_type, diversity_loss_type, feature_weights
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:37]: do you need the `self.loss weights` ? try to include only those that are needed later.

Comment (amit-sharma) [reply dice_ml/explainer_interfaces/dice_KD.py:37]: same for `optimizer_weights`

File: dice_ml/explainer_interfaces/dice_KD.py (hunk 1 above)
Comment (amit-sharma) [reply dice_ml/explainer_interfaces/dice_KD.py:43]: in fact, both lines can be folded (43 and 44) into a single function `get_num_output_nodes()` inside `self.model`, which does exactly these two operations.

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        self.feature_weights_input = ''
+        self.hyperparameters = [1, 1, 1]  # proximity_weight, diversity_weight, categorical_penalty
+        self.optimizer_weights = []  # optimizer, learning_rate
+
+        # number of output nodes of ML model
+        temp_input = tf.convert_to_tensor([tf.random.uniform([len(self.data_interface.encoded_feature_names)])], dtype=tf.float32)
+        self.num_ouput_nodes = self.model.get_output(temp_input).shape[1]
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:44]: typo: output, not ouput

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+        # Stores the predictions on the training data
+        #TODO: Implement this for private dataset instance too
+        dataset_dict = self.data_interface.prepare_dataset_instance(encode=True)
+        dataset_dict_output = np.array([dataset_dict.values], dtype=np.float32)
+
+        self.predictions = self.predict_fn(dataset_dict_output)
+        self.predictions_vals = np.reshape(self.predictions[0], (self.predictions.shape[1],))
+
+        #segmenting the dataset according to outcome
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:54]: init is usually only for initialization. the code to build kd trees (after line 54) should ideally come under a separate function, `build_kd_trees` that you simply call from __init__.

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        self.dataset_with_predictions[0] = dataset_with_predictions_all.loc[np.round(dataset_with_predictions_all[self.data_interface.outcome_name + "_pred"])==0]
+
+        # Prepares the KD trees for DiCE - 1 for each outcome (here only 0 and 1)
+        #TODO: make separate KD trees for each outcome
+        self.KD_tree = {
+            1: KDTree(pd.get_dummies(self.dataset_with_predictions[1][self.data_interface.feature_names])),
+            0: KDTree(pd.get_dummies(self.dataset_with_predictions[0][self.data_interface.feature_names]))}
+
+    def generate_counterfactuals(self, query_instance, total_CFs, desired_class="opposite", proximity_weight=0.5, diversity_weight=1.0, categorical_penalty=0.1, algorithm="DiverseCF", features_to_vary="all", permitted_range=None, yloss_type="hinge_loss", diversity_loss_type="dpp_style:inverse_dist", feature_weights="inverse_mad", optimizer="tensorflow:adam", learning_rate=0.05, min_iter=500, max_iter=5000, project_iter=0, loss_diff_thres=1e-5, loss_converge_maxiter=1, verbose=False, init_near_query_instance=True, tie_random=False, stopping_threshold=0.5, posthoc_sparsity_param=0.1, posthoc_sparsity_algorithm="linear"):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:67]: among the optional parameters, only include the parameters that you need. Remove all else.

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+        :return: A CounterfactualExamples object to store and visualize the resulting counterfactual explanations (see diverse_counterfactuals.py).
+
+        """
+
+        # check feature MAD validity and throw warnings
+        if feature_weights == "inverse_mad":
+            self.data_interface.get_valid_mads(display_warnings=True, return_mads=False)
+
+        # check permitted range for continuous features
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:107]: code for checking permitted range should move to a base class explainerbase, since it can apply to all methods and we want to avoid repeating code. do you want to take a go at it?

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+        return exp.CounterfactualExamples(self.data_interface, query_instance,
+        test_pred, self.final_cfs, self.cfs_preds, self.final_cfs_sparse, self.cfs_preds_sparse, posthoc_sparsity_param, desired_class)
+
+    def predict_fn(self, input_instance):
+        """prediction function"""
+        temp_preds = self.model.get_output(input_instance).numpy()
+        return np.array([preds[(self.num_ouput_nodes-1):] for preds in temp_preds], dtype=np.float32)
+
+    def do_cf_initializations(self, total_CFs, algorithm, features_to_vary):
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:133]: again this should be part of explainer base, since it applies to all methods.

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+
+        #preparing query instance for conversion to pandas dataframe
+        for q in query_instance_copy:
+            query_instance_copy[q] = [query_instance_copy[q]]
+        query_instance_df = pd.DataFrame.from_dict(query_instance_copy)
+
+        start_time = timeit.default_timer()
+
+        #Making the one-hot-encoded version of query instance match the one-hot encoded version of the dataset
+        query_instance_df_dummies = pd.get_dummies(query_instance_df)
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:186]: why is line 84 needed? this also makes think that the one-hot encoding code here and in DiCE should be moved to a separate function, so that other functions can call it easily.  It could be a static function in explainer_base.

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+                query_instance_df_dummies[col] = 0
+
+        # Finding counterfactuals from the KD Tree
+        indices = self.KD_tree[desired_class].query(query_instance_df_dummies, self.total_CFs)[1][0].tolist()
+        final_cfs = self.dataset_with_predictions[desired_class][self.data_interface.feature_names].iloc[indices].copy()
+
+        #finding the predicted outcome for each cf
+        self.cfs_preds = []
+        for i in range(len(indices)):
+            cfs = final_cfs.iloc[i].copy()
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:198]: again, the code leads to conversion from df->dict->df. not needed, try to directly work with the df. the prepare_query_instance assumed dict input from the user, that's why its input is a dict.

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+            cfs_dict = self.data_interface.prepare_query_instance(query_instance=cfs_dict, encode=True)
+            cfs_dict = np.array([cfs_dict.iloc[0].values])
+            test_pred_cf = self.predict_fn(cfs_dict)
+            test_pred_cf = np.reshape(test_pred_cf[0], (test_pred_cf.shape[1],))
+
+            self.cfs_preds.append(test_pred_cf)
+
+        #normalizing cfs here because we un-normalize in diverse_counterfactuals.py
+        #TODO: skip this step after making relevant changes in diverse_counterfactuals.py
+        final_cfs = self.data_interface.normalize_data(final_cfs)
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:209]: yeah, need to think what is the correct way here

File: dice_ml/explainer_interfaces/dice_KD.py [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+        # Making the one-hot-encoded version of final cfs match the one-hot encoded version of the dataset
+        final_cfs_dummies = pd.get_dummies(final_cfs)
+        for col in self.data_interface.encoded_feature_names:
+            if col not in final_cfs_dummies.columns:
```
Comment (amit-sharma) [inline dice_ml/explainer_interfaces/dice_KD.py:214]: do not understand why this is `if` condition is needed.

**Review by amit-sharma — COMMENTED** [review]
This is a good start. Added my comments inline
